service/src/main/kotlin/app/cash/backfila/service/runner/statemachine/BatchAwaiter.kt

```diff
@@ -183,14 +184,44 @@ class BatchAwaiter(
         ),
       )
 
-      // If all states are COMPLETE the whole backfill will be completed.
+      // If all states are COMPLETE, for every other partition, the whole backfill will be completed.
+      // In this case, don't mark the partition as complete until after the finalize call. This
+      // will ensure that the BackfillRunner is not marked as NOT running can cancel this coroutine.
       // If multiple partitions finish at the same time they will retry due to the hibernate
       // version mismatch on the DbBackfillRun.
       val partitions = dbRunPartition.backfill_run.partitions(
         session,
         backfillRunner.factory.queryFactory,
       )
-      if (partitions.all { it.run_state == BackfillState.COMPLETE }) {
+      partitions
```

Comment:
we should do these server side changes in a separate PR, and get them deployed before making the client side changes. Otherwise, a client developer might implement the finalize method when it doesn't do anything yet.

Also, there are problems with this code I mentioned in the previous PR. I will see if I can take a look at writing this correctly in the next few days

Reply:
@shellderp just curious if any progress was made here?

Reply:
I know how to code it up, read the version and retry both transactions if the 2nd read a different version. I can do this in a followup if you get the code merged as is (just add tests in this PR)